src/Std/Data/DTreeMap/Basic.lean

```diff
@@ -0,0 +1,355 @@
+-/
+@[inline]
+def getD (t : DTreeMap α β cmp) (a : α) (fallback : β) : β :=
+  letI : Ord α := ⟨cmp⟩; Impl.Const.getD a t.inner fallback
+
+end Const
+
+universe w w₂
+variable {δ : Type w} {m : Type w → Type w₂} [Monad m]
+
+/-- Removes all mappings of the map for which the given function returns `false`. -/
+@[inline]
+def filter (f : (a : α) → β a → Bool) (t : DTreeMap α β cmp) : DTreeMap α β cmp :=
+  letI : Ord α := ⟨cmp⟩; ⟨t.inner.filter f t.wf.balanced |>.impl, t.wf.filter⟩
+
+/--
+Folds the given monadic function over the mappings in the map in ascending order.
+-/
+@[inline]
+def foldlM (f : δ → (a : α) → β a → m δ) (init : δ) (t : DTreeMap α β cmp) : m δ :=
+  t.inner.foldlM f init
+
+/--
+Folds the given function over the mappings in the map in ascending order.
+-/
+@[inline]
+def foldl (f : δ → (a : α) → β a → δ) (init : δ) (t : DTreeMap α β cmp) : δ :=
+  t.inner.foldl f init
+
+/-- Carries out a monadic action on each mapping in the tree map in ascending order. -/
+@[inline]
+def forM (f : (a : α) → β a → m PUnit) (t : DTreeMap α β cmp) : m PUnit :=
+  t.inner.forM f
+
+/-- Support for the `for` loop construct in `do` blocks. Iteration happens in ascending order. -/
+@[inline]
+def forIn (f : (a : α) → β a → δ → m (ForInStep δ)) (init : δ) (t : DTreeMap α β cmp) : m δ :=
+  t.inner.forIn (fun c a b => f a b c) init
+
+instance : ForM m (DTreeMap α β cmp) ((a : α) × β a) where
+  forM t f := t.forM (fun a b => f ⟨a, b⟩)
+
+instance : ForIn m (DTreeMap α β cmp) ((a : α) × β a) where
+  forIn m init f := m.forIn (fun a b acc => f ⟨a, b⟩ acc) init
+
+/-- Check if any element satisfes the predicate, short-circuiting if a predicate fails. -/
+@[inline]
+def any (t : DTreeMap α β cmp) (p : (a : α) → β a → Bool) : Bool := Id.run $ do
+  for ⟨a, b⟩ in t do
+    if p a b then return true
+  return false
+
+/-- Check if all elements satisfy the predicate, short-circuiting if a predicate fails. -/
+@[inline]
+def all (t : DTreeMap α β cmp) (p : (a : α) → β a → Bool) : Bool := Id.run $ do
+  for ⟨a, b⟩ in t do
+    if p a b = false then return false
+  return true
+
+/-- Returns a list of all keys present in the tree map in some order. -/
+@[inline]
+def keys (t : DTreeMap α β cmp) : List α :=
+  t.inner.keys
+
+/-- Returns an array of all keys present in the tree map in some order. -/
+@[inline]
+def keysArray (t : DTreeMap α β cmp) : Array α :=
+  t.inner.keysArray
+
+/-- Transforms the tree map into a list of mappings in ascending order. -/
+@[inline]
+def toList (t : DTreeMap α β cmp) : List ((a : α) × β a) :=
+  t.inner.toList
+
+/-- Transforms the tree map into a list of mappings in ascending order. -/
+@[inline]
+def toArray (t : DTreeMap α β cmp) : Array ((a : α) × β a) :=
+  t.inner.toArray
+
+/--
+Returns a map that contains all mappings of `t₁` and `t₂`. In case that both maps contain the
+same key `k` with respect to `cmp`, the provided function is used to determine the new value from
+the respective values in `t₁` and `t₂`.
```

Comment:
Now that you put so much work into this function, perhaps add a sentence about linearity, similarly as it exists for `alter`/`modify` on the hash map?

Reply:
I added explanations of the sense in which this function is linear and of the performance characteristics.